### DISCUSSION TITLE
Mapping AuxVariable that is Monomial to a Lagrangian one....

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#26626] [Q&A General]
I've been playing with a problem where I have a populated monomial variable, I want to populate a material object with the contents of that variable. What is the right way to do this? I basically am I trying to use the gravity kernel from tensor mechanics, which requires density as a material object, my material density is currently variable contained within a monomial auxvariable.

**Comment by GiudGiud** [comment]
Hello

The ParsedMaterial can convert a variable into a material property.
If you want to use AD (and density depends on a nonlinear variable), rewriting your density as a material property is the way forward instead of forwarding from an auxiliary variable

Guillaume

**Reply by makeclean** [reply]
So, what im doing is trying to use topology optimisation on 'real' geometry where some bits can be modified and some can't. So for example, I have two blocks, the red (not being optimised) and white being optimised. With the toy problems i've fixed the ends, and I'm pulling down (gravity nodal kernel) on the bottom. 

![image](https://github.com/idaholab/moose/assets/2439754/df2195ce-1e6c-4f72-8543-2cd7c2b84c77)

However, I want to apply it to a real problem, and thus I'm building complexity by setting the density to be fixed in the red, and variable on the basis of the SIMP method. I would like to use the tensor mechanics gravity kernel to determine the gravitational forces. 

The examples found in the optimisation folder of the installation, have  AuxVariables like

```[AuxVariables]
  [Emin]
    family = MONOMIAL
    order = CONSTANT
    initial_condition = ${Emin}
    block = 'inner'
  []
  [power]
    family = MONOMIAL
    order = CONSTANT
    initial_condition = ${power}
    block = 'inner'      
  []
  [E0]
    family = MONOMIAL
    order = CONSTANT
    initial_condition = ${E0}
    block = 'inner'     
  []
  [Dc]
    family = MONOMIAL
    order = CONSTANT
    initial_condition = -1.0
    block = 'inner'
  []
  [mat_den]
    family = MONOMIAL
    order = CONSTANT
  []
```
So I was using the density here, and setting it block with;

```
[ICs]
  [block1]
    type = ConstantIC
    variable = mat_den
    value = ${vol_frac}
    block = 'inner'
  []
  [block2]
    type = ConstantIC
    variable = mat_den
    value = 1.0
    block = 'outer'
  []
[]
```
I was then filling the nodal AuxVariable with the density, and then using
```
[Materials]
  [density]
    type = GenericFunctionMaterial
    prop_names = 'density'
    prop_values = get_density
  []
[]

[Functions]
  [get_density]
    type = ParsedFunction
    expression = 'alpha'
    symbol_names = 'alpha'
    symbol_values = mat_den_nodal
  []
[]
```
This segfaults however; 
```
*** ERROR ***
Unsupported order: 0
Stack frames: 18
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&)
2: /home/adavis/opt/moose/moose/scripts/../libmesh/installed/lib/libmesh_opt.so.0(+0x787861) [0x7f1feb387861]
3: libMesh::DofMap::reinit(libMesh::MeshBase&)
4: libMesh::DofMap::distribute_dofs(libMesh::MeshBase&)
5: libMesh::System::init_data()
6: libMesh::System::init()
7: libMesh::EquationSystems::init()
8: FEProblemBase::init()
9: Action::timedAct()
10: ActionWarehouse::executeActionsWithAction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
11: ActionWarehouse::executeAllActions()
12: MooseApp::runInputFile()
13: MooseApp::run()
14: main
15: /lib64/libc.so.6(+0x27550) [0x7f1fe284a550]
16: __libc_start_main
17: /home/adavis/opt/moose/moose/modules/combined/combined-opt() [0x4044f5]
[0] ../src/fe/fe_lagrange.C, line 777, compiled Nov  7 2023 at 15:03:29
```

**Reply by makeclean** [reply]
So instead, you're saying I should set density via the material object, and instead use an AuxKernel to populate the AuxVariable, so MaterialRealAux to populate the AuxVariable. That would give the ability to set whats needed, but at the end of the timestep I would need to update that material object, would that be a PostProcessor?

**Reply by GiudGiud** [reply]
what is `mat_den_nodal` ?
```
[Functions]
  [get_density]
    type = ParsedFunction
    expression = 'alpha'
    symbol_names = 'alpha'
    symbol_values = mat_den_nodal
  []
[]
```
functions cant depend on variables. 

The seg fault is concerning. We definitely would want your input or a reduced version of it to fix this. 

You can use the `MaterialRealAux ` to populate the auxvariable indeed. The variable is only needed for output right? The material property is actually what is being used in the optimization ?

**Reply by makeclean** [reply]
No, the material property isn't being used in the optimisation, I've modified https://github.com/idaholab/moose/blob/next/modules/combined/examples/optimization/2d_mbb.i to use the Gravity kernel, but the the density, mat_den is a monomial auxvariable since the element is either density greater than 0 (on) and 0 (off), that density is updated by a postprocessor on the basis of the SIMP compliance. I was trying to update to use the kernel, such that this is reflect in variable mass rather than a fixed mass in that example.

**Reply by GiudGiud** [reply]
@lynnmunday

**Reply by GiudGiud** [reply]
@somu15

**Reply by lynnmunday** [reply]
@maxnezdyur Has examples that limit the optimization to a specific block while not optimizing the other blocks for a fixed load.  Is the load you are applying a function of the density so as the optimization changes the density field, it also changes the load?  I don't know that I've seen an example of SIMP for a problem like that.  @maxnezdyur will know, he implemented all of this.
